observability/opentelemetry/v2/client/otel_observability_service.go

```diff
@@ -0,0 +1,197 @@
+/*
+ Copyright 2021 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package client
+
+import (
+	"context"
+	"strings"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.opentelemetry.io/otel/trace"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/observability"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+	cehttp "github.com/cloudevents/sdk-go/v2/protocol/http"
+)
+
+// OTelObservabilityService implements the ObservabilityService interface from cloudevents
+type OTelObservabilityService struct {
+	tracer               trace.Tracer
+	spanAttributesGetter func(cloudevents.Event) []attribute.KeyValue
+	spanNameFormatter    func(cloudevents.Event) string
+}
+
+// NewOTelObservabilityService returns a OpenTelemetry enabled observability service
+func NewOTelObservabilityService(opts ...OTelObservabilityServiceOption) *OTelObservabilityService {
+	tracerProvider := otel.GetTracerProvider()
+
+	os := &OTelObservabilityService{
+		tracer: tracerProvider.Tracer(
+			instrumentationName,
+			trace.WithInstrumentationVersion("1.0.0"), // TODO: Can we have the package version here?
+		),
+		spanNameFormatter: defaultSpanNameFormatter,
+	}
+
+	// apply passed options
+	for _, opt := range opts {
+		opt(os)
+	}
+
+	return os
+}
+
+// InboundContextDecorators returns a decorator function that allows enriching the context with the incoming parent trace.
+// This method gets invoked automatically by passing the option 'WithObservabilityService' when creating the cloudevents HTTP client.
+func (os OTelObservabilityService) InboundContextDecorators() []func(context.Context, binding.Message) context.Context {
+	return []func(context.Context, binding.Message) context.Context{tracePropagatorContextDecorator}
+}
+
+// RecordReceivedMalformedEvent records the error from a malformed event in the span.
+func (os OTelObservabilityService) RecordReceivedMalformedEvent(ctx context.Context, err error) {
+	spanName := observability.ClientSpanName + ".malformed receive"
+	_, span := os.tracer.Start(
+		ctx, spanName,
+		trace.WithSpanKind(trace.SpanKindConsumer),
+		trace.WithAttributes(attribute.String(string(semconv.CodeFunctionKey), "RecordReceivedMalformedEvent")))
+
+	recordSpanError(span, err)
+	span.End()
+}
+
+// RecordCallingInvoker starts a new span before calling the invoker upon a received event.
+// In case the operation fails, the error is recorded and the span is marked as failed.
+func (os OTelObservabilityService) RecordCallingInvoker(ctx context.Context, event *cloudevents.Event) (context.Context, func(errOrResult error)) {
+	spanName := os.getSpanName(event, "receive")
```

Comment:
```suggestion
	spanName := os.getSpanName(event, "process")
```
This is not tracking busy waiting for a message but the handler function processing it, right?
See https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md.

Reply:
Yeah this is invoked when the busy-waiting process receives the event, right before calling the handler function. 

With the client you get from cloudevents, you can listen (blocking) like this: https://github.com/cloudevents/sdk-go#receive-your-first-cloudevent. 

Then this method on the obs service is called from the invoker: https://github.com/cloudevents/sdk-go/blob/release-2.5/v2/client/invoker.go#L83.

I think both operation names are "appropriate" so I'm not sure which one is more semantic.

Reply:
I changed to `process`. I agree it makes more sense.